src/targets/node/nodeAttacherBase.ts

```diff
@@ -0,0 +1,89 @@
+/*---------------------------------------------------------
+ * Copyright (C) Microsoft Corporation. All rights reserved.
+ *--------------------------------------------------------*/
+
+import { AnyNodeConfiguration } from '../../configuration';
+import { NodeLauncherBase, IRunData, IProcessTelemetry } from './nodeLauncherBase';
+import { IProgram } from './program';
+import Cdp from '../../cdp/api';
+import { INodeTargetLifecycleHooks } from './nodeTarget';
+import { IDisposable } from '../../common/disposable';
+import { logger } from '../../common/logging/logger';
+import { LogTag } from '../../common/logging';
+
+/**
+ * Base class that implements common matters for attachment.
+ */
+export abstract class NodeAttacherBase<T extends AnyNodeConfiguration> extends NodeLauncherBase<T> {
+  /**
+   * Tracker for whether we're waiting to break and instrument into the main
+   * process. This is used to avoid instrumenting unecessarily into subsequent
+   * children.
+   */
+  private capturedEntryProgram?: IProgram;
+
+  /**
+   * @inheritdoc
+   */
+  protected createLifecycle(cdp: Cdp.Api, run: IRunData<T>): INodeTargetLifecycleHooks {
+    if (this.program === this.capturedEntryProgram) {
+      return {};
+    }
+
+    let toDispose: Promise<ReadonlyArray<IDisposable>> = Promise.resolve([]);
+    this.capturedEntryProgram = this.program;
+
+    return {
+      initialized: async () => {
+        toDispose = this.onFirstInitialize(cdp, run);
+        await toDispose;
```

Comment:
We set it like this so that in the event a close happens before initialization finishes, we don't leak resources